apps/ex_wire/lib/ex_wire/struct/warp_queue.ex

```diff
@@ -0,0 +1,275 @@
+    :unprocessed_block_data
+  ]
+
+  @doc """
+  Creates a new `WarpQueue` that 
+  """
+  def new() do
+    %__MODULE__{
+      manifests: [],
+      first_block_hash: nil,
+      current_block_number: nil,
+      current_block_hash: nil,
+      known_block_hashes: MapSet.new(),
+      known_state_hashes: MapSet.new(),
+      block_chunk_requests: MapSet.new(),
+      unprocessed_block_data: %{}
+    }
+  end
+
+  @spec new_manifest(t(), SnapshotManifest.manifest()) :: t()
+  def new_manifest(warp_queue, manifest) do
+    updated_known_block_hashes =
+      MapSet.union(
+        warp_queue.known_block_hashes,
+        MapSet.new(manifest.block_hashes)
+      )
+
+    updated_known_state_hashes =
+      MapSet.union(
+        warp_queue.known_state_hashes,
+        MapSet.new(manifest.state_hashes)
+      )
+
+    %{
+      warp_queue
+      | manifests: [manifest | warp_queue.manifests],
+        first_block_hash: List.first(manifest.block_hashes),
+        current_block_number: nil,
+        current_block_hash: nil,
+        known_block_hashes: updated_known_block_hashes,
+        known_state_hashes: updated_known_state_hashes
+    }
+  end
+
+  @doc """
+  When we receive a new block chunk, we want to remove it from requests
+  and add it to our processing queue.
+  """
+  @spec new_block_chunk(WarpQueue.t(), BlockChunk.t()) :: WarpQueue.t()
+  def new_block_chunk(warp_queue, block_chunk) do
+    next_warp_queue =
+      if !is_nil(warp_queue.first_block_hash) &&
+           ExthCrypto.Hash.Keccak.kec(ExRLP.encode(BlockChunk.serialize(block_chunk))) ==
+             warp_queue.first_block_hash do
+        %{
+          warp_queue
+          | first_block_hash: nil,
+            current_block_number: block_chunk.number,
+            current_block_hash: block_chunk.hash
+        }
+      else
+        warp_queue
+      end
+
+    updated_block_chunk_requests =
+      MapSet.delete(next_warp_queue.block_chunk_requests, block_chunk.hash)
+
+    {_, updated_unprocessed_block_data} =
+      Enum.reduce(
+        block_chunk.block_data_list,
+        {block_chunk.number, next_warp_queue.unprocessed_block_data},
+        fn block_data, {number, unprocessed_block_data} ->
+          {number + 1, Map.put(unprocessed_block_data, number, block_data)}
+        end
+      )
+
+    Exth.inspect(
+      [next_warp_queue.current_block_number, Map.keys(updated_unprocessed_block_data)],
+      "unprocessed block numbers"
+    )
+
+    %{
+      next_warp_queue
+      | block_chunk_requests: updated_block_chunk_requests,
+        unprocessed_block_data: updated_unprocessed_block_data
+    }
+  end
+
+  # TODO: We need to handle requesting state hashes, as well.
+  @spec get_block_hashes_to_request(t(), integer()) :: {t(), list(EVM.hash())}
+  def get_block_hashes_to_request(warp_queue, request_limit) do
+    blocks_to_request = request_limit - Enum.count(warp_queue.block_chunk_requests)
+
+    if blocks_to_request > 0 do
+      unrequested_blocks =
```

Comment:
I think you could pipe this into `blocks_hashes_to_request` i.e.

```ex
blocks_hashes_to_request =
  warp_queue.known_block_hashes
  |> MapSet.difference(warp_queue.block_chunk_requests)
  |> Enum.take(blocks_to_request)
```

Reply:
You're right, will do.